### DISCUSSION TITLE
Single coalton-toplevel vs multiple coalton-toplevel

### DISCUSSION THREAD

**Discussion by digikar99** [coalton-lang/coalton#993] [Help!]
What difference does having a single coalton-toplevel block make over having multiple coalton-toplevel blocks? Say, in the following example, what will be the difference?

```lisp
(coalton-toplevel
  (define-type (pair :a) (make-pair :a :a))
  (define (bar x)
    (match x
      ((make-pair a b)
       (tuple a b)))))
```

```lisp
(coalton-toplevel
  (define-type (pair :a) (make-pair :a :a)))

(coalton-toplevel
  (define (bar x)
    (match x
      ((make-pair a b)
       (tuple a b)))))
```

If there's little difference, then I can imagine writing a helper package/system for coalton that provides shadowed coalton-toplevel operators that expand into a `(coalton-toplevel (top-level-operator ...))` and interact with coalton the way I use common lisp:

```
(cl:defmacro coalton-helper:define (var-or-fun &body body)
  `(coalton-toplevel
     (define ,var-or-fun ,@body)))
```

**Comment by eliaslfox** [comment]
You can't have recursive definitions between multiple coalton toplevels.

**Reply by digikar99** [reply]
> shadowed coalton-toplevel operators

I don't mean they will expand recursively. They will just "look like" coalton toplevel operators aka their symbol names will be the same. On the implementation side, they won't actually be toplevel operators, but will be implemented as common lisp macros that expand into a `(coalton-toplevel (actual-toplevel-operator ...))` form.
